compiler/src/dotty/tools/dotc/core/JavaNullInterop.scala

```diff
@@ -0,0 +1,160 @@
+    new JavaNullMap(false)(ctx)(tp)
+
+  /** A type map that implements the nullification function on types. Given a Java-sourced type, this adds `| JavaNull`
+   *  in the right places to make the nulls explicit in Scala.
+   *
+   *  @param outermostLevelAlreadyNullable whether this type is already nullable at the outermost level.
+   *                                       For example, `Array[String]|JavaNull` is already nullable at the
+   *                                       outermost level, but `Array[String|JavaNull]` isn't.
+   *                                       If this parameter is set to true, then the types of fields, and the return
+   *                                       types of methods will not be nullified.
+   *                                       This is useful for e.g. constructors, and also so that `A & B` is nullified
+   *                                       to `(A & B) | JavaNull`, instead of `(A|JavaNull & B|JavaNull) | JavaNull`.
+   */
+  private class JavaNullMap(var outermostLevelAlreadyNullable: Boolean)(implicit ctx: Context) extends TypeMap {
+    /** Should we nullify `tp` at the outermost level? */
+    def needsNull(tp: Type): Boolean =
+      !outermostLevelAlreadyNullable && (tp match {
+        case tp: TypeRef =>
+          // We don't modify value types because they're non-nullable even in Java.
+          !tp.symbol.isValueClass &&
+          // We don't modify `Any` because it's already nullable.
+          !tp.isRef(defn.AnyClass) &&
+          // We don't nullify Java varargs at the top level.
+          // Example: if `setNames` is a Java method with signature `void setNames(String... names)`,
+          // then its Scala signature will be `def setNames(names: (String|JavaNull)*): Unit`.
+          // This is because `setNames(null)` passes as argument a single-element array containing the value `null`,
+          // and not a `null` array.
+          !tp.isRef(defn.RepeatedParamClass)
+        case _ => true
+      })
+
+    /** Should we nullify the type arguments to the given generic `tp`?
+     *  We only nullify the inside of Scala-defined generics.
+     *  This is because Java classes are _all_ nullified, so both `java.util.List[String]` and
+     *  `java.util.List[String|Null]` contain nullable elements.
+     */
+    def needsNullArgs(tp: AppliedType): Boolean = !tp.classSymbol.is(JavaDefined)
+
+    override def apply(tp: Type): Type = {
+      // Fast version of Type::toJavaNullableUnion that doesn't check whether the type
+      // is already a union.
+      def toJavaNullableUnion(tpe: Type): Type = OrType(tpe, defn.JavaNullAliasType)
+
+      tp match {
+        case tp: TypeRef if needsNull(tp) => toJavaNullableUnion(tp)
+        case appTp @ AppliedType(tycons, targs) =>
+          val oldOutermostNullable = outermostLevelAlreadyNullable
+          outermostLevelAlreadyNullable = false
+          val targs2 = if (needsNullArgs(appTp)) targs map this else targs
```

Comment:
What if the bounds of the type param don't admit `Null`?  E.g. `class MyBox[T <: Boxable]` --> `MyBox[T | JavaNull]` is not well-formed.

Reply:
This is a good question. Let me get back to you on this: we will probably need to nullify some upper/lower bounds that show up in Java wildcards.